core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleManager.scala

```diff
@@ -79,9 +81,9 @@ private[spark] class SortShuffleManager(conf: SparkConf) extends ShuffleManager
   }
 
   /**
-   * A mapping from shuffle ids to the number of mappers producing output for those shuffles.
+   * A mapping from shuffle ids to the task ids of mappers producing output for those shuffles.
    */
-  private[this] val numMapsForShuffle = new ConcurrentHashMap[Int, Int]()
+  private[this] val taskIdMapsForShuffle = new ConcurrentHashMap[Int, ArrayBuffer[Long]]()
```

Comment:
After using the map task attempt id as part of shuffle file, here we must record all the map task attempt id for each shuffle id. So comparing with the original implement, it's a memory waste here. But consider the shuffle map task number, it's maybe an accessible change?

Reply:
mostly it's hundreds or thousands of mappers, which is about several KB. I think it's fine. If there are many many mappers, our DAG scheduler will burn out first.

Reply:
Yeah.

Reply:
Well, @xuanyuanking should already be aware of that 100K Mappers is not that rare for large production jobs. That would be ~10MB for single one map stage.

Maybe we should removes old shuffleId's data just like the scheduler removes old stages. However I do believe it's fine for now. Let's revise this when it actually hits.

Reply:
Yeah that's why I highlight this question here, just because we see the huge job before. :)
For current implement of ContextCleaner, the in-memory shuffle metrics bind with JVM gc, the config `spark.cleaner.periodicGC.interval` can help us.
Sure, let's keep tracking.

Reply:
Thanks for the guidance, change it to use `OpenHashSet`, and gives a smaller initial size `16`.